ch08-3Dgraphics/shapesandmesh/src/main/java/org/modernclientjava/javafx3d/ShapesAndMesh.java

```diff
@@ -0,0 +1,216 @@
+        }
+
+        public final ObjectProperty<CullFace> cullFaceProperty() {
+            return cullFace;
+        }
+    }
+
+    private static class View {
+
+        public Scene scene;
+
+        public Sphere sphere;
+        public Cylinder cylinder;
+        public Box box;
+        public MeshView meshView;
+
+        public ComboBox<DrawMode> drawModeComboBox;
+        public ComboBox<CullFace> cullFaceComboBox;
+        public Slider rotateSlider;
+
+        public View(Model model) {
+            sphere = new Sphere(50);
+            cylinder = new Cylinder(50, 100);
+            box = new Box(100, 100, 100);
+            meshView = createMeshView(100);
+
+            sphere.setTranslateX(-300);
+            cylinder.setTranslateX(-100);
+            box.setTranslateX(100);
+            meshView.setTranslateX(300);
+
+            sphere.setMaterial(new PhongMaterial(Color.RED));
+            cylinder.setMaterial(new PhongMaterial(Color.YELLOW));
+            box.setMaterial(new PhongMaterial(Color.BLUE));
+            meshView.setMaterial(new PhongMaterial(Color.GREEN));
+
+            setupShape3D(sphere, model);
+            setupShape3D(cylinder, model);
+            setupShape3D(box, model);
+            setupShape3D(meshView, model);
+
+            Group shapesGroup = new Group(sphere, cylinder, box, meshView);
+
+            drawModeComboBox = new ComboBox<>();
+            drawModeComboBox.setItems(FXCollections.observableArrayList(
+                    DrawMode.FILL, DrawMode.LINE));
+            drawModeComboBox.setValue(DrawMode.FILL);
+
+            cullFaceComboBox = new ComboBox<>();
+            cullFaceComboBox.setItems(FXCollections.observableArrayList(
+                    CullFace.BACK, CullFace.FRONT, CullFace.NONE));
+            cullFaceComboBox.setValue(CullFace.BACK);
+
+            HBox hbox1 = new HBox(10, new Label("DrawMode:"), drawModeComboBox,
+                    new Label("CullFace:"), cullFaceComboBox);
+            hbox1.setPadding(new Insets(10, 10, 10, 10));
+            hbox1.setAlignment(Pos.CENTER_LEFT);
+
+            rotateSlider = new Slider(-180.0d, 180.0d, 60.0d);
+            rotateSlider.setMinWidth(400.0d);
+            rotateSlider.setMajorTickUnit(10.0d);
+            rotateSlider.setMinorTickCount(5);
+            rotateSlider.setShowTickMarks(true);
+            rotateSlider.setShowTickLabels(true);
+
+            rotateSlider.valueProperty().bindBidirectional(model.rotateProperty());
+
+            HBox hbox2 = new HBox(10, new Label("Rotate Around (1, 1, 1) Axis:"),
+                    rotateSlider);
+            hbox2.setPadding(new Insets(10, 10, 10, 10));
+            hbox2.setAlignment(Pos.CENTER_LEFT);
+
+            VBox controlPanel = new VBox(10, hbox1, hbox2);
+            controlPanel.setPadding(new Insets(10, 10, 10, 10));
+
+            BorderPane root = new BorderPane(shapesGroup, null, null, controlPanel, null);
+            scene = new Scene(root, 800, 600);
```

Comment:
add Antialiasing; 
`scene = new Scene(root, 800, 600, true, SceneAntialiasing.BALANCED);`
This means that now the shapes have to be displayed in a `SubScene`, with the following changes:


```
import javafx.scene.SceneAntialiasing;
import javafx.scene.SubScene;
....
            sphere.setTranslateX(100);
            cylinder.setTranslateX(300);
            box.setTranslateX(500);
            meshView.setTranslateX(700);

...
            Group shapesGroup = new Group(sphere, cylinder, box, meshView);
            SubScene subScene = new SubScene(shapesGroup, 800, 400, true, SceneAntialiasing.BALANCED);

...
            Group groupSubScene = new Group(subScene);
            BorderPane root = new BorderPane(groupSubScene, null, null, controlPanel, null);
            scene = new Scene(root, 800, 600, true, SceneAntialiasing.BALANCED);
```

Reply:
This is a good use case to show the combination of 2D and 3D in the same scene, that requires the use of a `SubScene` (I recall adding a comment about this in the chapter)